Solve stdc++ ABI incompatibility between manylinux nupic.bindings and recent Linux distros

It turns out that the c\+\+11 standard prescribed a new ABI for certain classes, but compiler toolchains took their own sweet time to implement that ABI. While they support -std=c\+\+11, the implementation of the new ABI is not uniform. The baseline C++ toolchain prescribed by the manylinux policy supports std=c\+\+11, but doesn’t implement the new ABI. The 5.4.0 toolchain on Ubuntu 16.04 implements the new ABI, which alters declarations of some standard classes, and exceptions in particular.

One area where this impacts us is in the ability to catch some exceptions. For example, https://github.com/numenta/nupic.core/blob/b9247a270e2fae2f4586286163b6472a909560e2/src/nupic/regions/VectorFile.cpp#L177 catches ios_base::failure&. This is one of the exceptions impacted by the ABI change. In the manylinux toolchain, it’s simply derived from std::exception, which is the signature that gets compiled into nupic.core when building a manylinux wheel. However, the new ABI mandates that it should be derived from std::system_error (and sets some additional attributes), which gives it a different signature. Thus, our manylinux build is unable to catch that exception when running on a system whose stdc\+\+ library incorporates the new ABI (e.g. Ubuntu 16.04). There is a discussion about it in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66145.

For this specific case, the workaround may be to catch std::exception instead of or in addition to ios_base::failure or to restructure the use of streaming API to not make use of exceptions. However, the vast body of 3rd party code in nupic.core also makes use of this specific exception,  and the ABI change may impact some other classes, likely making it impractical to make such adaptations in the overall vast code base.

Another possibility that may be more practical:
1. Include libstdc++ statically
2. And use a version script in the builds of the nupic.bindings extension shared libs that hides all symbols except the extension initialization functions.
 * When libstdc\+\+ links statically, it's symbol references still show up as public in the generated extension shared libs. This means that if something else in the application (e.g., pycapnp) causes the shared libstdc\+\+ to be loaded, then the C\+\+ references in our extensions might be linked to the shared libstdc\+\+ that implements the new ABI instead of the statically-linked libstdc\+\+.

```
integration/nupic/engine/vector_file_sensor_test.py:85: VectorFileSensorTest.testAll FAILED

============================================================== FAILURES ==============================================================
____________________________________________________ VectorFileSensorTest.testAll ____________________________________________________

self = <tests.integration.nupic.engine.vector_file_sensor_test.VectorFileSensorTest testMethod=testAll>

    def testAll(self):
      """Run all the tests in our suite, catching any exceptions that might be
        thrown.
        """
      print 'VectorFileSensorTest parameters:'
      print 'PYTHONPATH: %s' % os.environ.get('PYTHONPATH', 'NOT SET')
      print 'filename: %s' % self.filename
    
      self._testRunWithoutFile()
      self._testNetLoad()
    
      self._testFakeLoadFile()
      self._testRepeatCount()
      self._testUnknownCommand()
    
      # Test maxOutput and activeOutputCount
      self._testOutputCounts(0)
>     self._testLoadFile(self.dataFile, '0', '0')

integration/nupic/engine/vector_file_sensor_test.py:102: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.integration.nupic.engine.vector_file_sensor_test.VectorFileSensorTest testMethod=testAll>
dataFile = '/home/vitaly/nta/nupic/tests/integration/nupic/engine/data/vectortestdata.txt', fileFormat = '0', iterations = '0'

    def _testLoadFile(self, dataFile, fileFormat= '', iterations=''):
      """Test reading our sample vector file. The sample file
        has 5 vectors of the correct length, plus one with incorrect length.
        The sensor should ignore the last line."""
    
      # Now load a real file
      if fileFormat != '':
>       res = self.sensor.executeCommand(['loadFile', dataFile, fileFormat])

integration/nupic/engine/vector_file_sensor_test.py:208: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <nupic.engine.Region object at 0x7f9d27b6e110>
args = ['loadFile', '/home/vitaly/nta/nupic/tests/integration/nupic/engine/data/vectortestdata.txt', '0']

    def executeCommand(self, args):
      """
        @doc:place_holder(Region.executeCommand)
        """
>     return self._region.executeCommand(args)

../../../.local/lib/python2.7/site-packages/nupic/engine/__init__.py:492: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <nupic.bindings.engine_internal.Region; proxy of <Swig Object of type 'nupic::Region *' at 0x7f9d2a0df7e0> >
args = (['loadFile', '/home/vitaly/nta/nupic/tests/integration/nupic/engine/data/vectortestdata.txt', '0'],), kwargs = {}

    def executeCommand(self, *args, **kwargs):
        """executeCommand(self, args) -> std::string"""
>       return _engine_internal.Region_executeCommand(self, *args, **kwargs)
E       RuntimeError: basic_ios::clear: iostream error

../../../.local/lib/python2.7/site-packages/nupic/bindings/engine_internal.py:1431: RuntimeError
---------------------------------------------------------- Captured stdout -----------------------------------------------------------
VectorFileSensorTest parameters:
PYTHONPATH: NOT SET
filename: /home/vitaly/nta/nupic/tests/integration/nupic/engine/data/vectorfile.nta
In VectorFileSensor ExistenceIsAnIllusion.txt
ERR:  VectorFile::appendFile - unable to open file: ExistenceIsAnIllusion.txt [/nupic.core/src/nupic/regions/VectorFile.cpp line 105]
ERR:  VectorFileSensor: Unknown execute command: 'nonExistentCommand' sent! [/nupic.core/src/nupic/regions/VectorFileSensor.cpp line 320]
In VectorFileSensor /home/vitaly/nta/nupic/tests/integration/nupic/engine/data/vectortestdata.txt
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================== 1 failed, 70 passed, 3 skipped in 90.94 seconds ===========================================




(gdb) bt
#0  nupic::VectorFile::appendFile (this=this@entry=0x23eba88, fileName="data/vectortestdata.txt", expectedElementCount=11, fileFormat=fileFormat@entry=0)
    at /nupic.core/src/nupic/regions/VectorFile.cpp:110
#1  0x00007f89ad242aac in nupic::VectorFileSensor::executeCommand (this=0x23eba60, args=..., index=<optimized out>)
    at /nupic.core/src/nupic/regions/VectorFileSensor.cpp:260
#2  0x00007f89ad1e003f in nupic::Region::executeCommand (this=0x23ebcc0, args=std::__debug::vector of length 3, capacity 4 = {...})
    at /nupic.core/src/nupic/engine/Region.cpp:295
#3  0x00007f89ad1594a0 in _wrap_Region_executeCommand (args=<optimized out>, kwargs=<optimized out>)
    at /nupic.core/build/scripts/src/nupic/bindings/engine_internalPYTHON_wrap.cxx:31361
#4  0x00000000004cb26d in PyEval_EvalFrameEx ()


Manylinux on ubuntu 16.04:
vitaly@ubuntuvm:~/nta/nupic/tests/integration/nupic/engine$ ldd ~/.local/lib/python2.7/site-packages/nupic/bindings/_engine_internal.so 
	linux-vdso.so.1 =>  (0x00007ffc9eb55000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fcfcb83d000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fcfcb639000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fcfcb2b6000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcfcafad000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fcfcad97000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcfca9cd000)
	/lib64/ld-linux-x86-64.so.2 (0x00005643a9082000)
	
Local build on ubuntu 16.04
vitaly@ubuntuvm:~/nta/nupic/tests/integration/nupic/engine$ ldd ~/.local/lib/python2.7/site-packages/nupic/bindings/_engine_internal.so 
	linux-vdso.so.1 =>  (0x00007ffddb054000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f6f97ff6000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f6f97de0000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6f97a16000)
	/lib64/ld-linux-x86-64.so.2 (0x000056372bf21000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6f9770d000)

Manylinux on centos 6.8
[root@1979a0512c63 cp27-cp27mu]# ldd /opt/python/cp27-cp27mu/lib/python2.7/site-packages/nupic/bindings/_engine_internal.so 
	linux-vdso.so.1 =>  (0x00007ffda87ca000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f76eb4b1000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f76eb2ac000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f76eafa6000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f76ead22000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f76eab0b000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f76ea777000)
	/lib64/ld-linux-x86-64.so.2 (0x0000555dddbb4000)
```

Here are the areas in the g++ 5.4.0 headers that are affected by the new ABI:

```
/usr/include/c++/5.4.0$ grep -r _GLIBCXX_USE_CXX11_ABI .
./bits/locale_facets_nonio.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/locale_facets_nonio.h:      && _GLIBCXX_USE_CXX11_ABI == 0
./bits/locale_facets_nonio.h:      && _GLIBCXX_USE_CXX11_ABI == 0
./bits/locale_facets_nonio.h:      && _GLIBCXX_USE_CXX11_ABI == 0
./bits/locale_facets_nonio.h:      && _GLIBCXX_USE_CXX11_ABI == 0
./bits/locale_classes.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/basic_string.tcc:#if _GLIBCXX_USE_CXX11_ABI
./bits/basic_string.tcc:#else  // !_GLIBCXX_USE_CXX11_ABI
./bits/basic_string.tcc:#endif  // !_GLIBCXX_USE_CXX11_ABI
./bits/stl_list.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/stl_list.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/ios_base.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/ios_base.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/locale_facets_nonio.tcc:      && _GLIBCXX_USE_CXX11_ABI == 0
./bits/locale_facets_nonio.tcc:      && _GLIBCXX_USE_CXX11_ABI == 0
./bits/basic_string.h:#if _GLIBCXX_USE_CXX11_ABI
./bits/basic_string.h:#else  // !_GLIBCXX_USE_CXX11_ABI
./bits/basic_string.h:#endif  // !_GLIBCXX_USE_CXX11_ABI
./sstream:#if _GLIBCXX_USE_CXX11_ABI
./stdexcept:#if _GLIBCXX_USE_CXX11_ABI
./stdexcept:#else // _GLIBCXX_USE_CXX11_ABI
./stdexcept:#endif // _GLIBCXX_USE_CXX11_ABI
./stdexcept:#if _GLIBCXX_USE_CXX11_ABI || _GLIBCXX_DEFINE_STDEXCEPT_COPY_OPS
./stdexcept:#if _GLIBCXX_USE_CXX11_ABI || _GLIBCXX_DEFINE_STDEXCEPT_COPY_OPS
./system_error:#if _GLIBCXX_USE_CXX11_ABI
./experimental/fs_fwd.h:#if _GLIBCXX_USE_CXX11_ABI
```

--- a/CommonCompilerConfig.cmake
+++ b/CommonCompilerConfig.cmake
@@ -183,12 +183,17 @@ endif()
 
 if (${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU")
   if (${NUPIC_BUILD_PYEXT_MODULES} AND "${PLATFORM}" STREQUAL "linux")
-    # For python extensions on Linux, we want shared libgcc and
-    # libstdc++ in order to avoid memory management issues and other
-    # side-effects of static versions of those libs on various distros.
-    # NOTE There is no `-shared-libstdc++` flag; shared libstdc++ is given
-    # priority over its static counterpart implicitly when `-static-libstdc++`
-    # is omitted.
+    # NOTE When building manylinux python extensions, we want the static
+    # libstdc++ due to differences in c++ ABI between the older toolchain in the
+    # manylinux Docker image and libstdc++ in newer linux distros that is
+    # compiled with the c++11 ABI. for example, with shared libstdc++, the
+    # manylinux-built extension is unable to catch std::ios::failure exception
+    # raised by the shared libstdc++.so while running on Ubuntu 16.04.
+    set(stdlib_cxx "${stdlib_cxx} -static-libstdc++")
+
+    # NOTE We need to use shared libgcc to be able to throw and catch exceptions
+    # across different shared libraries, as may be the case when our python
+    # extensions runtime-link to capnproto symbols in pycapnp's extension.
     set(stdlib_common "${stdlib_common} -shared-libgcc")
   else()
     set(stdlib_common "${stdlib_common} -static-libgcc")

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -690,6 +690,7 @@ if (NUPIC_BUILD_PYEXT_MODULES)
 
   # Common dependencies for our python extensions for use with
   # SWIG_MODULE_name_EXTRA_DEPS
+  # Make sure we don't execute the swig executable before it is built
   set(_SRC_SWIG_EXTRA_DEPS Swig)
 
   # NOTE Windows DLLs are shared executables with their own main; they require
@@ -713,19 +714,19 @@ if (NUPIC_BUILD_PYEXT_MODULES)
 
 
   function(PREPEND_BOILERPLATE_TO_PYTHON_PROXY_MODULE
-           PARENT_SWIG_TARGET)
-    # Add a custom command to the given Swig target to prepend boilerplate to
-    # the swig-generated python proxy module
-    # ${CMAKE_SWIG_OUTDIR}/${PARENT_SWIG_TARGET}.py. The boilerplate preloads
+           MODULE_NAME)
+    # Add a custom command to the Swig target to prepend boilerplate to the
+    # swig-generated python proxy module
+    # ${CMAKE_SWIG_OUTDIR}/${MODULE_NAME}.py. The boilerplate preloads
     # the pycapnp extension shared library on Unix platforms.
     #
-    # :param PARENT_SWIG_TARGET: the custom command will be added to this
-    #   swig_add_module target
+    # :param MODULE_NAME: the custom command will be added to the target
+    #   corresponding to this Swig module name.
     set(preamble_filepath "${CMAKE_SOURCE_DIR}/src/nupic/bindings/swig_proxy_preamble.py")
-    set(module_filepath "${CMAKE_SWIG_OUTDIR}/${PARENT_SWIG_TARGET}.py")
+    set(module_filepath "${CMAKE_SWIG_OUTDIR}/${MODULE_NAME}.py")
 
     add_custom_command(
-      TARGET ${SWIG_MODULE_${PARENT_SWIG_TARGET}_REAL_NAME}
+      TARGET ${SWIG_MODULE_${MODULE_NAME}_REAL_NAME}
       POST_BUILD
       COMMAND
         ${CMAKE_COMMAND}
@@ -738,48 +739,85 @@ if (NUPIC_BUILD_PYEXT_MODULES)
   endfunction(PREPEND_BOILERPLATE_TO_PYTHON_PROXY_MODULE)
 
 
-  function(BUILD_EXTENSION
-           EXTENSION_NAME)
-    # Create a nupic.bindings swig extension target with the given python
-    # extension name. Also, if PY_EXTENSIONS_DIR is specified, request
+  function(BUILD_EXTENSION MODULE_NAME)
+    # Create a nupic.bindings swig extension target with the given Swig module
+    # name. Also, if PY_EXTENSIONS_DIR is specified, request
     # installation of the extension library and python proxy module.
     #
-    # The real target name is ${SWIG_MODULE_${EXTENSION_NAME}_REAL_NAME}.
+    # The real target name is ${SWIG_MODULE_${MODULE_NAME}_REAL_NAME}.
 
-    set(source_interface_file nupic/bindings/${EXTENSION_NAME}.i)
+    set(source_interface_file nupic/bindings/${MODULE_NAME}.i)
 
     set_source_files_properties(${source_interface_file} PROPERTIES
                                 CPLUSPLUS ON
-                                SWIG_MODULE_NAME ${EXTENSION_NAME})
+                                SWIG_MODULE_NAME ${MODULE_NAME})
 
-    # Make sure we don't execute the swig executable before it is built
-    set(SWIG_MODULE_${EXTENSION_NAME}_EXTRA_DEPS ${_SRC_SWIG_EXTRA_DEPS})
-
+    #
     # Create custom command for generating files from SWIG
+    #
 
     # Note: swig_add_module outputs ${swig_generated_file_fullname}
-    swig_add_module(${EXTENSION_NAME} python ${source_interface_file})
-    swig_link_libraries(${EXTENSION_NAME}
-                        ${_SRC_SWIG_LINK_LIBRARIES})
-    set_target_properties(${SWIG_MODULE_${EXTENSION_NAME}_REAL_NAME} PROPERTIES
-                          LINK_FLAGS "${_SRC_SWIG_EXTENSION_LINK_FLAGS}")
-    prepend_boilerplate_to_python_proxy_module(${EXTENSION_NAME})
+    swig_add_module(${MODULE_NAME} python ${source_interface_file})
 
-    # Set properties on swig-generated files
     set_source_files_properties(
       ${swig_generated_file_fullname} PROPERTIES
       GENERATED TRUE
       COMPILE_FLAGS ${src_swig_generated_file_compile_flags})
 
+    swig_link_libraries(${MODULE_NAME} ${_SRC_SWIG_LINK_LIBRARIES})
+
+    prepend_boilerplate_to_python_proxy_module(${MODULE_NAME})
+
+    set(real_target "${SWIG_MODULE_${MODULE_NAME}_REAL_NAME}")
+
+    set(extra_deps ${_SRC_SWIG_EXTRA_DEPS})
+    set(link_flags ${_SRC_SWIG_EXTENSION_LINK_FLAGS})
+
+    # Create an export map and update extra dependencies and link flags. This
+    # export map prevents runtime-link preemption of statically-linked
+    # libraries, such as -static-libstdc++, and limits the shared object's
+    # symbol visibility to only the python extension's init function. NOTE Not
+    # sure what, if anything, to do for MSVC at this time.
+    set(extension_init_func "init${real_target}")
+
+    if("${PLATFORM}" STREQUAL "darwin")
+      set(link_flags
+          "${link_flags} -Wl,-exported_symbol,_${extension_init_func}")
+
+    elseif("${PLATFORM}" STREQUAL "linux" OR MINGW)
+      set(export_map_file
+          "${CMAKE_CURRENT_BINARY_DIR}/${real_target}_generated.expmap")
+
+      list(APPEND extra_deps "${export_map_file}")
+      set(link_flags "${link_flags} -Wl,--version-script=${export_map_file}")
+
+      set(export_map_contents "{global: ${extension_init_func}; local: *;};")
+
+      message(STATUS "Writing export map file ${export_map_file} "
+              "(${export_map_contents}).")
+
+      file(WRITE ${export_map_file} "${export_map_contents}")
+    endif()
+
+
+    set(SWIG_MODULE_${MODULE_NAME}_EXTRA_DEPS ${extra_deps})
+    set_target_properties(${real_target} PROPERTIES
+                          LINK_FLAGS "${link_flags}")
+
     # If a path is specified, copy extensions files to proper location.
     if (PY_EXTENSIONS_DIR)
       install(TARGETS
-              ${SWIG_MODULE_${EXTENSION_NAME}_REAL_NAME}
+              ${real_target}
               LIBRARY DESTINATION ${PY_EXTENSIONS_DIR})
       install(FILES
-              ${PROJECT_BINARY_DIR}/${EXTENSION_NAME}.py
+              ${PROJECT_BINARY_DIR}/${MODULE_NAME}.py
               DESTINATION ${PY_EXTENSIONS_DIR})
     endif(PY_EXTENSIONS_DIR)
+
+    message(
+      STATUS
+      "Created Swig target ${real_target} for swig module ${MODULE_NAME}. "
+      "extra_deps=${extra_deps}, link_flags=${link_flags}")
   endfunction(BUILD_EXTENSION)
 
 

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -421,9 +421,9 @@ set(src_external_static_libs
 # Add capnproto static lib when building nupic_core static lib without python
 # extensions. The extensions have their own logic governing whether to include
 # it.
-if (NOT ${NUPIC_BUILD_PYEXT_MODULES})
+if (NOT NUPIC_BUILD_PYEXT_MODULES)
   list(APPEND src_external_static_libs ${CAPNP_STATIC_LIB_TARGET})
-endif() # ${NUPIC_BUILD_PYEXT_MODULES}
+endif()
 
 set(src_combined_nupiccore_source_archives
     ${src_lib_static_nupiccore_solo}
@@ -447,9 +447,9 @@ set(src_common_test_exe_libs
 # Add capnproto static lib to our C/C++ test apps when building python
 # extensions, since capnproto objects are excluded from nupic_core combined lib
 # to avoid conflict with capnproto methods compiled into the pycapnp extension
-if (${NUPIC_BUILD_PYEXT_MODULES})
+if (NUPIC_BUILD_PYEXT_MODULES)
   list(APPEND src_common_test_exe_libs ${CAPNP_STATIC_LIB_TARGET})
-endif() # ${NUPIC_BUILD_PYEXT_MODULES}
+endif()
 
 
 #
@@ -620,7 +620,7 @@ add_custom_target(tests_all
 #
 # Use SWIG to generate Python extensions.
 #
-if (${NUPIC_BUILD_PYEXT_MODULES})
+if (NUPIC_BUILD_PYEXT_MODULES)
   include(UseSWIG)
 
   message(STATUS "CMAKE_CXX_COMPILER_ID        = ${CMAKE_CXX_COMPILER_ID}")
@@ -671,7 +671,7 @@ if (${NUPIC_BUILD_PYEXT_MODULES})
   set(CMAKE_SWIG_FLAGS ${src_swig_flags} ${CMAKE_SWIG_FLAGS})
 
   # Set up linker flags for python extension shared libraries
-  set(src_swig_extension_link_flags "${PYEXT_LINKER_FLAGS_OPTIMIZED}")
+  set(_SRC_SWIG_EXTENSION_LINK_FLAGS "${PYEXT_LINKER_FLAGS_OPTIMIZED}")
 
   # NOTE Non-Windows Python extensions shouldn't be linking agains libpython;
   # symbols should be available automatically when python loads the extension.
@@ -684,13 +684,13 @@ if (${NUPIC_BUILD_PYEXT_MODULES})
   # capnproto symbols by preloading the pycapnp extension in our extension
   # python proxy modules, thus avoiding the conflict from executing the methods
   # in our own compilation of capnproto on objects created by pycapnp's.
-  set(src_swig_link_libraries
+  set(_SRC_SWIG_LINK_LIBRARIES
       ${src_lib_static_nupiccore_combined}
       ${src_common_os_libs})
 
   # Common dependencies for our python extensions for use with
   # SWIG_MODULE_name_EXTRA_DEPS
-  set(src_swig_extra_deps Swig)
+  set(_SRC_SWIG_EXTRA_DEPS Swig)
 
   # NOTE Windows DLLs are shared executables with their own main; they require
   # all symbols to resolve at link time, so we have to add libpython for this
@@ -701,14 +701,14 @@ if (${NUPIC_BUILD_PYEXT_MODULES})
   # nupic/nupic.bindings presently have no dependency on pycapnp. We also include
   # PYTHON_LIBRARIES because all symbols have to resolve when building a DLL.
   if("${PLATFORM}" STREQUAL "windows")
-    list(APPEND src_swig_link_libraries
+    list(APPEND _SRC_SWIG_LINK_LIBRARIES
          ${PYTHON_LIBRARIES}
          ${CAPNP_STATIC_LIB_TARGET})
   endif()
 
-  message(STATUS "src_swig_extra_deps     = ${src_swig_extra_deps}")
-  message(STATUS "src_swig_link_libraries = ${src_swig_link_libraries}")
-  message(STATUS "src_swig_extension_link_flags= ${src_swig_extension_link_flags}")
+  message(STATUS "_SRC_SWIG_EXTRA_DEPS     = ${_SRC_SWIG_EXTRA_DEPS}")
+  message(STATUS "_SRC_SWIG_LINK_LIBRARIES = ${_SRC_SWIG_LINK_LIBRARIES}")
+  message(STATUS "_SRC_SWIG_EXTENSION_LINK_FLAGS= ${_SRC_SWIG_EXTENSION_LINK_FLAGS}")
   message(STATUS "CMAKE_SWIG_FLAGS        = ${CMAKE_SWIG_FLAGS}")
 
 
@@ -738,100 +738,64 @@ if (${NUPIC_BUILD_PYEXT_MODULES})
   endfunction(PREPEND_BOILERPLATE_TO_PYTHON_PROXY_MODULE)
 
 
+  function(BUILD_EXTENSION
+           EXTENSION_NAME)
+    # Create a nupic.bindings swig extension target with the given python
+    # extension name. Also, if PY_EXTENSIONS_DIR is specified, request
+    # installation of the extension library and python proxy module.
+    #
+    # The real target name is ${SWIG_MODULE_${EXTENSION_NAME}_REAL_NAME}.
+
+    set(source_interface_file nupic/bindings/${EXTENSION_NAME}.i)
+
+    set_source_files_properties(${source_interface_file} PROPERTIES
+                                CPLUSPLUS ON
+                                SWIG_MODULE_NAME ${EXTENSION_NAME})
+
+    # Make sure we don't execute the swig executable before it is built
+    set(SWIG_MODULE_${EXTENSION_NAME}_EXTRA_DEPS ${_SRC_SWIG_EXTRA_DEPS})
+
+    # Create custom command for generating files from SWIG
+
+    # Note: swig_add_module outputs ${swig_generated_file_fullname}
+    swig_add_module(${EXTENSION_NAME} python ${source_interface_file})
+    swig_link_libraries(${EXTENSION_NAME}
+                        ${_SRC_SWIG_LINK_LIBRARIES})
+    set_target_properties(${SWIG_MODULE_${EXTENSION_NAME}_REAL_NAME} PROPERTIES
+                          LINK_FLAGS "${_SRC_SWIG_EXTENSION_LINK_FLAGS}")
+    prepend_boilerplate_to_python_proxy_module(${EXTENSION_NAME})
+
+    # Set properties on swig-generated files
+    set_source_files_properties(
+      ${swig_generated_file_fullname} PROPERTIES
+      GENERATED TRUE
+      COMPILE_FLAGS ${src_swig_generated_file_compile_flags})
+
+    # If a path is specified, copy extensions files to proper location.
+    if (PY_EXTENSIONS_DIR)
+      install(TARGETS
+              ${SWIG_MODULE_${EXTENSION_NAME}_REAL_NAME}
+              LIBRARY DESTINATION ${PY_EXTENSIONS_DIR})
+      install(FILES
+              ${PROJECT_BINARY_DIR}/${EXTENSION_NAME}.py
+              DESTINATION ${PY_EXTENSIONS_DIR})
+    endif(PY_EXTENSIONS_DIR)
+  endfunction(BUILD_EXTENSION)
+
+
   # Algorithms
-  set(src_swig_algorithms_files nupic/bindings/algorithms.i)
-  set_source_files_properties(${src_swig_algorithms_files} PROPERTIES
-                              CPLUSPLUS ON
-                              SWIG_MODULE_NAME algorithms)
-  set(src_swig_algorithms algorithms)
-  # Make sure we don't attempt to execute the swig executable until it is built.
-  set(SWIG_MODULE_${src_swig_algorithms}_EXTRA_DEPS ${src_swig_extra_deps})
-  # Create custom command for generating files from SWIG.
-  swig_add_module(${src_swig_algorithms} python ${src_swig_algorithms_files})
-  list(APPEND src_swig_generated_files ${swig_generated_file_fullname})
-  swig_link_libraries(${src_swig_algorithms}
-                      ${src_swig_link_libraries})
-  set_target_properties(${SWIG_MODULE_${src_swig_algorithms}_REAL_NAME} PROPERTIES
-                        LINK_FLAGS "${src_swig_extension_link_flags}")
-  prepend_boilerplate_to_python_proxy_module(${src_swig_algorithms})
+  build_extension("algorithms")
 
   # Engine
-  set(src_swig_engine_files nupic/bindings/engine_internal.i)
-  set_source_files_properties(${src_swig_engine_files} PROPERTIES
-                              CPLUSPLUS ON
-                              SWIG_MODULE_NAME engine_internal)
-  set(src_swig_engine engine_internal)
-  # Make sure we don't attempt to execute the swig executable until it is built.
-  set(SWIG_MODULE_${src_swig_engine}_EXTRA_DEPS ${src_swig_extra_deps})
-  # Create custom command for generating files from SWIG.
-  swig_add_module(${src_swig_engine} python ${src_swig_engine_files})
-  list(APPEND src_swig_generated_files ${swig_generated_file_fullname})
-  swig_link_libraries(${src_swig_engine}
-                      ${src_swig_link_libraries})
-  set_target_properties(${SWIG_MODULE_${src_swig_engine}_REAL_NAME} PROPERTIES
-                        LINK_FLAGS "${src_swig_extension_link_flags}")
-  prepend_boilerplate_to_python_proxy_module(${src_swig_engine})
+  build_extension("engine_internal")
 
   # Experimental
-  set(src_swig_experimental_files nupic/bindings/experimental.i)
-  set_source_files_properties(${src_swig_experimental_files} PROPERTIES
-                              CPLUSPLUS ON
-                              SWIG_MODULE_NAME experimental)
-  set(src_swig_experimental experimental)
-  # Make sure we don't attempt to execute the swig executable until it is built.
-  set(SWIG_MODULE_${src_swig_experimental}_EXTRA_DEPS ${src_swig_extra_deps})
-  # Create custom command for generating files from SWIG.
-  swig_add_module(${src_swig_experimental} python ${src_swig_experimental_files})
-  list(APPEND src_swig_generated_files ${swig_generated_file_fullname})
-  swig_link_libraries(${src_swig_experimental}
-                      ${src_swig_link_libraries})
-  set_target_properties(${SWIG_MODULE_${src_swig_experimental}_REAL_NAME} PROPERTIES
-                        LINK_FLAGS "${src_swig_extension_link_flags}")
-  prepend_boilerplate_to_python_proxy_module(${src_swig_experimental})
+  build_extension("experimental")
 
   # Math
-  set(src_swig_math_files nupic/bindings/math.i)
-  set_source_files_properties(${src_swig_math_files} PROPERTIES
-                              CPLUSPLUS ON
-                              SWIG_MODULE_NAME math)
-  set(src_swig_math math)
-  # Make sure we don't attempt to execute the swig executable until it is built.
-  set(SWIG_MODULE_${src_swig_math}_EXTRA_DEPS ${src_swig_extra_deps})
-  # Create custom command for generating files from SWIG.
-  swig_add_module(${src_swig_math} python ${src_swig_math_files})
-  list(APPEND src_swig_generated_files ${swig_generated_file_fullname})
-  swig_link_libraries(${src_swig_math}
-                      ${src_swig_link_libraries})
-  set_target_properties(${SWIG_MODULE_${src_swig_math}_REAL_NAME} PROPERTIES
-                        LINK_FLAGS "${src_swig_extension_link_flags}")
-  prepend_boilerplate_to_python_proxy_module(${src_swig_math})
+  build_extension("math")
 
-
-  # Set properties on swig-generated and support files
-  set_source_files_properties(${src_swig_generated_files} PROPERTIES
-                              GENERATED TRUE)
-  set_source_files_properties(
-    ${src_swig_generated_files}
-    PROPERTIES
-    COMPILE_FLAGS ${src_swig_generated_file_compile_flags})
-
-  # If a path is specified, copy extensions files to proper location.
-  if (PY_EXTENSIONS_DIR)
-    install(TARGETS
-            ${SWIG_MODULE_${src_swig_algorithms}_REAL_NAME}
-            ${SWIG_MODULE_${src_swig_engine}_REAL_NAME}
-            ${SWIG_MODULE_${src_swig_experimental}_REAL_NAME}
-            ${SWIG_MODULE_${src_swig_math}_REAL_NAME}
-            LIBRARY DESTINATION ${PY_EXTENSIONS_DIR})
-    install(FILES
-            ${PROJECT_BINARY_DIR}/algorithms.py
-            ${PROJECT_BINARY_DIR}/engine_internal.py
-            ${PROJECT_BINARY_DIR}/experimental.py
-            ${PROJECT_BINARY_DIR}/math.py
-            DESTINATION ${PY_EXTENSIONS_DIR})
-  endif(PY_EXTENSIONS_DIR)
-
-endif() # ${NUPIC_BUILD_PYEXT_MODULES}
+endif() # NUPIC_BUILD_PYEXT_MODULES
 
 
 #

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -699,8 +699,7 @@ if (NUPIC_BUILD_PYEXT_MODULES)
   #
   # NOTE On Windows nupic.bindings builds, we include capnproto because we
   # presently build self-contained CAPNP_LITE on Windows, and Windows
-  # nupic/nupic.bindings presently have no dependency on pycapnp. We also include
-  # PYTHON_LIBRARIES because all symbols have to resolve when building a DLL.
+  # nupic/nupic.bindings presently have no dependency on pycapnp.
   if("${PLATFORM}" STREQUAL "windows")
     list(APPEND _SRC_SWIG_LINK_LIBRARIES
          ${PYTHON_LIBRARIES}

